### ISSUE TITLE
how to extract yolov5 features to add other deeplearning alogrithm

### ISSUE THREAD

**Issue by besbesmany** [ultralytics/yolov5#4095]
## ❔Question
I want to add any other deeplearning technique like ssd , vgg16 after yolov5

how to take features from yolov5 and merge them with other deeplearning technique

can you help me please or send me link or code for this?

**Comment by trungpham2606** [comment]
Dear @glenn-jocher,
I have the same question with @besbesmany 
Iam trying to add mask branch after yolov5. But Iam struggle in this part, my question is how can I get the xywh of each roi in this feature map here
![image](https://user-images.githubusercontent.com/35657426/126478616-cfa21061-0177-4dea-aff9-b03fc5f55b23.png)
If I understand correctly, z contains the xywh coordinates corresponding to image's resolution. But I want the xywh coordinates corresponding to the each feature map x(i). 
How can I get them ?
Btw, can you explain the formula here:
![image](https://user-images.githubusercontent.com/35657426/126478965-968087f5-295c-4d3f-9c0f-5551fd2e6479.png)

**Comment by glenn-jocher** [comment]
@trungpham2606 if you want to add a mask output you probably want to start from the backbone end and essentially make an inverted backbone to bring the features up to a higher resolution.

The feature maps into the Detect() layer are very small, i.e. at stride 32 a 640x640 image would be 20x20, so probably not suitable for segmentation.

**Comment by trungpham2606** [comment]
Dear @glenn-jocher 
Actually the feature maps into the Detect() layer is a list of 3 features map (320x32,x32, 640x16x16, 1280x8x8)
The output feature of maskrcnn is also (14x14x256), so I think it's not the priority problem now.
I want to make sure that in the formula below:
![image](https://user-images.githubusercontent.com/35657426/126615507-80872da4-91ff-4064-8adf-c58829f96d79.png)
after going through the Detect() layer, we can convert the output coordinate back to image's coordinate by the above formula.
So what is the xywh in the output feature map coordinates ? Can you make it clear for me ? 
According to the formula, if we dont multiply self.stride[i], the y[...:0:2] will be the x,y coordinate related to the output feature map, but for y[..., 2:4], if I ignore the term * self.anchor_grid[i], the y[..., 2:4] will be the wh coor related to output feature map ?

**Comment by glenn-jocher** [comment]
@trungpham2606 the lines you show create boxes in image coordinates (0-640). If you want them in grid coordinates (0-20) just divide by stride.

**Comment by trungpham2606** [comment]
@glenn-jocher thank you for making it clear for me. 
I have added the mask head after the backbone model and trained with yolo, but seems there is something I have to change in order to make it work well.

**Comment by glenn-jocher** [comment]
@trungpham2606 that's interesting. What does your loss function look like? Is this instance segmentation or semantic segmentation?

**Comment by trungpham2606** [comment]
@glenn-jocher 
It's aspired by the maskrcnn, the difference thing is I used yolo for output ROI. So it's instance segmentation. The loss Iam using for mask part is binary_cross_entropy.
I had added mask head to some other 1 stage objects detection such as ssd and centernet. The mask head's structure is similar to the original maskrcnn's mask head. The results are pretty good (although I trained and tested on my dataset only).
There are now some new works for instance segmentation that can be integrated to 1 stage objects detection.

**Comment by glenn-jocher** [comment]
@trungpham2606 do you think you might be able to submit a PR for segmentation? It's a request we hear about from our users but I don't have a lot of segmentation experience myself.

In the segmentation loss if you have say 10 objects and a batch-size 1 image of shape 1x3x640x640 that reduces to say 1x1024x20x20 at the head what shape would the segmentation loss be before it reduces?

**Comment by trungpham2606** [comment]
@glenn-jocher 
I changed your code so much, just made it as much simple as possible (training strategies, augmentation, ...). Beside, I'm still not able to get good results.
If we have 10 objects or something, just calculate the loss for each objects and find the average, that's what I did. This is my simple mask loss function:
![image](https://user-images.githubusercontent.com/35657426/126644196-ce14bcee-7753-41c2-a12f-7bf1a830aef0.png)

**Comment by besbesmany** [comment]
@trungpham2606 can you help me 
what should I do in code to extract yolov5 features to add it to another deeplearning technique vgg16 or ssd?

**Comment by trungpham2606** [comment]
@besbesmany 
The feature maps I think you are asking for are before feeding to Detect layer. Iam also taking them as output feature maps.
So you can change the code like this
![image](https://user-images.githubusercontent.com/35657426/126854288-9b60ce25-229c-40f5-a0e8-8c02ab22dc56.png)
So x_features is a list which contains 3 output feature maps.

**Comment by trungpham2606** [comment]
@glenn-jocher @besbesmany 
These are some of my yolo instance segmentation's results. It has a big room to improve the results, this is my very first achievement (all objects belong to 1 same class):
![image](https://user-images.githubusercontent.com/35657426/126857449-87a9c451-8920-42b4-bb77-c1c403cd5d34.png)
![image](https://user-images.githubusercontent.com/35657426/126857454-f16c76cf-3599-4be9-8000-5c67cf55a7ea.png)
![image](https://user-images.githubusercontent.com/35657426/126857464-a6547128-5c44-4a90-a2d4-2bb8121ac2f3.png)
![image](https://user-images.githubusercontent.com/35657426/126857473-8f977784-8950-4e66-8645-7197fefa8f29.png)

**Comment by glenn-jocher** [comment]
@trungpham2606 hey it works! Could you submit a PR so we could see the updates you made for segmentation? Thanks!

**Comment by trungpham2606** [comment]
@glenn-jocher 
I have made a PR.

**Comment by trungpham2606** [comment]
@glenn-jocher 
The changes I made are applying for batch-size 1 and 1 class training.
If you guys want to change the batch size and multi-class training, there're some places in code have to be changed.
This PR is like a simple pipeline how to combine mask head to yolo in order to solve instance segmentation tasks that other developers can follow.

**Comment by besbesmany** [comment]
@trungpham2606 can you send me modified files , I am still confused how to extract features and add another deeplearning methods

**Comment by glenn-jocher** [comment]
@besbesmany we don't provide assistance for custom code.

**Comment by trungpham2606** [comment]
@besbesmany 
I had pulled a request, you can check it in Pull Request section (yolo mask).
